File: catalog.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "catalog.h"


// Initialization Functions:

void initializeCatalog(Catalog* c, int count) {
    c -> tables = (TableSchema*)malloc(sizeof(TableSchema));
    c->tableCount = count;
}


// Insertion Functions:

void addTable(Catalog* c, TableSchema* table) {
    TableSchema* newTables = realloc(c->tables, (c->tableCount + 1) * sizeof(TableSchema));
    if(newTables != NULL) {
        newTables[c->tableCount] = *table;
        c->tables = newTables;
        c->tableCount++;
        printf("added table: %s\n", table->name);
    } else { fprintf(stderr, "Memory allocation failed\n"); }

}  // recordCount and currentPage initialized as 0

void dropTable(Catalog* c, char name[MAX_NAME_SIZE]) {
    int tableIndex = -1;

    for(int i = 0; i < c->tableCount; i++) {
        if(strcmp(c->tables[i].name, name) == 0) {
            tableIndex = i;
            break;
        }
    }

    if(tableIndex != -1) {
        for(int i = tableIndex; i < c->tableCount - 1; i++) {
            c->tables[i] = c->tables[i+1];
        }

        c->tableCount--;

        TableSchema* newTables = realloc(c->tables, c->tableCount * sizeof(TableSchema));

        if(newTables != NULL || c->tableCount == 0) {
            c->tables = newTables;
        } else {
            fprintf(stderr, "Table '%s' not found in catalog\n", name);
        }
    }
}

void addPage(Catalog* c) {

}  // How do we deal with this one?

void addAttribute(Catalog* c, char name[MAX_NAME_SIZE], char type[20]) {

}  // Add more constraints as needed

void displayCatalog(Catalog* c) {
    fprintf("Displaying %s", c->tableCount);
    for(int i = 0; i < c->tableCount; i++) {
        displayTable(&(c->tables[i]));
    }
    
    if(c->tableCount == 0) {
        printf("No such table\n");  // TODO: Do I need to include table name?
        printf("ERROR\n\n");
    }
}

void writeCatalogToFile(Catalog* c, char* pathname) {
    FILE* file = fopen(pathname, "wb");

    fwrite(c, sizeof(Catalog), 1, file);

    fwrite(c -> tables, sizeof(TableSchema), 1, file);

    fwrite(c -> tables -> attributes, sizeof(AttributeSchema), 1, file);

    fclose(file);
}

void readCatalogFromFile(Catalog* c, char* pathname) {
    FILE* file = fopen(pathname, "rb");

    fread(c, sizeof(Catalog), 1, file);

    c -> tables = (TableSchema*)malloc(sizeof(TableSchema));
    fread(c -> tables, sizeof(TableSchema), 1, file);

    c -> tables -> attributes = (AttributeSchema*)malloc(sizeof(AttributeSchema));
    fread(c -> tables -> attributes, sizeof(AttributeSchema), 1, file);

    fclose(file);
}

File: Catalog.java
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Catalog {
    private List<TableSchema> tables;
    private int tableCount;

    private String dbDirectory;
    private int pageSize;
    private int bufferSize;

    public Catalog(String dbDirectory, int pageSize, int bufferSize) {
        this.tables = new ArrayList<>();
        this.tableCount = 0;
        this.dbDirectory = dbDirectory;
        this.pageSize = pageSize;
        this.bufferSize = bufferSize;
    }

    public void addTable(TableSchema table) {
        if (table != null) {
            this.tables.add(table);
            this.tableCount++;
            System.out.println("Added table: " + table.getName());
        }
    }

    public void dropTable(String tableName) {
        boolean removed = this.tables.removeIf(table -> table.getName().equals(tableName));
        if (removed) {
            this.tableCount--;
            System.out.println("Table dropped: " + tableName);
        } else {
            System.err.println("Table not found: " + tableName);
        }
    }

    // TODO: Implement this
    public void addPage(Page newPage) {
        System.err.println("addPage method is dependent on system design and not directly implemented here.");
    }

    // TODO: Implement this
    public void addAttribute(String tableName, String attributeName, String type) {
        System.err.println("addAttribute method is dependent on system design and not directly implemented here.");
    }

    public void displayCatalog() {
        if (this.tables.isEmpty()) {
            System.out.println("No tables to display");
        } else {
            for (TableSchema table : this.tables) {
                System.out.println("Table: " + table.getName());
                table.displayTable();
            }
        }
    }

    public boolean tableExists(String name) {
        return this.tables.stream().anyMatch(table -> table.getName().equals(name));
    }

    public boolean findTableDisplay(String name) {
        TableSchema foundTable = this.tables.stream()
                                 .filter(table -> table.getName().equals(name))
                                 .findFirst()
                                 .orElse(null);
        if (foundTable != null) {
            foundTable.displayTable();
            return true;
        }
        return false;
    }

    public void writeCatalogToFile(String pathname) throws IOException {
        try (DataOutputStream dos = new DataOutputStream(new FileOutputStream(pathname))) {
            dos.writeInt(this.tableCount);
            for (TableSchema table : this.tables) {
                table.writeToStream(dos); 
            }
        }
    }

    public void readCatalogFromFile(String pathname) throws IOException {
        try (DataInputStream dis = new DataInputStream(new FileInputStream(pathname))) {
            this.tableCount = dis.readInt();
            this.tables.clear();
            for (int i = 0; i < this.tableCount; i++) {
                TableSchema table = TableSchema.readFromStream(dis); 
                this.tables.add(table);
            }
        }
    }

    public TableSchema getTableSchema(int tableNumber) {
        return this.tables.stream()
                .filter(table -> table.gettableNumber() == tableNumber)
                .findFirst()
                .orElse(null);
    }

    public int getNextPageNumber() {
        int maxPageNumber = -1;
        for (TableSchema table : this.tables) {
            int tableMaxPage = table.getMaxPageNumber();
            if (tableMaxPage > maxPageNumber) {
                maxPageNumber = tableMaxPage;
            }
        }
        return maxPageNumber + 1;
    }

    // TODO: Implement this
    public void updatePage(Page page) {
        System.err.println("updatePage method is dependent on system design and not directly implemented here.");
    }

    public int getTableCount() {
        return this.tableCount;
    }

    public List<TableSchema> getTables() {
        return this.tables;
    }

    public String getDbDirectory() {
        return this.dbDirectory;
    }

    public int getPageSize() {
        return this.pageSize;
    }

    public int getBufferSize() {
        return this.bufferSize;
    }

    public int getNextTableNumber() {
        return this.tables.size() + 1;
    }

    public TableSchema getTableSchemaByName(String tableName) {
        return this.tables.stream()
                .filter(table -> table.getName().equals(tableName))
                .findFirst()
                .orElse(null);
    }
}
